include/counter.h

```diff
@@ -22,35 +23,190 @@
 #include <zephyr/types.h>
 #include <stddef.h>
 #include <device.h>
+#include <stdbool.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef void (*counter_callback_t)(struct device *dev, void *user_data);
+/** @brief Alarm callback
+ *
+ * @param dev       Pointer to the device structure for the driver instance.
+ * @param chan_id   Channel ID.
+ * @param ticks     Counter value that triggered the callback.
+ * @param user_data User data.
+ */
+typedef void (*counter_alarm_callback_t)(struct device *dev,
+					 u8_t chan_id, u32_t ticks,
+					 void *user_data);
+
+/** @brief Alarm callback structure.
+ *
+ * @param callback Callback called on alarm (cannot be NULL).
+ * @param ticks Ticks that triggers the alarm. In case of absolute flag is set,
+ *		maximal value that can be set equals top value
+ *		(@ref counter_get_top_value). Otherwise
+ *		@ref counter_get_max_relative_alarm() returns maximal value that
+ *		can be set. If counter is clock driven then ticks can be
+ *		converted to microseconds (see @ref counter_ticks_to_us).
+ *		Alternatively, counter implementation may count asynchronous
+ *		events.
+ * @param user_data User data returned in callback.
+ * @param absolute Ticks relation to counter value. If true ticks are treated as
+ *		absolute value, else it is relative to the counter reading
+ *		performed during the call.
+ */
+struct counter_alarm_cfg {
+	counter_alarm_callback_t callback;
+	u32_t ticks;
+	void *user_data;
+	bool absolute;
+};
+
+/** @brief Callback called when counter turns around.
+ *
+ * @param dev       Pointer to the device structure for the driver instance.
+ * @param user_data User data provided in @ref counter_set_top_value.
+ */
+typedef void (*counter_top_callback_t)(struct device *dev, void *user_data);
+
+/** @brief Structure with generic counter features.
+ *
+ * @param max_top_value Maximal (default) top value on which counter is reset
+ *			(cleared or reloaded).
+ * @param freq		Frequency of the source clock if synchronous events are
+ *			counted.
+ * @param count_up	Flag indicating direction of the counter. If true
+ *			counter is counting up else counting down.
+ * @param channels	Number of channels that can be used for setting alarm,
+ *			see @ref counter_set_alarm.
+ */
+struct counter_config_info {
```

Comment:
Somewhere we need a comment about how this has to be the first element in the driver's config struct.